pkg/scheduler/admission/podnodeconstraints/admission.go

```diff
@@ -0,0 +1,175 @@
+	}
+	switch attr.GetResource() {
+	// We only want CREATE for pods
+	case kapi.Resource("pods"):
+		if attr.GetOperation() == admission.Update {
+			return nil
+		}
+	// But we want CREATE & UPDATE for every other resource
+	case kapi.Resource("replicationcontrollers"),
+		extensions.Resource("deployments"),
+		extensions.Resource("replicasets"),
+		extensions.Resource("jobs"),
+		deployapi.Resource("deploymentconfigs"):
+		// These are the only types we care about
+	default:
+		return nil
+	}
+	ps, err := o.getPodSpec(attr)
+	if err == nil {
+		return o.admitPodSpec(attr, ps)
+	}
+	return err
+}
+
+// extract the PodSpec from the pod templates for each object we care about
+func (o *podNodeConstraints) getPodSpec(attr admission.Attributes) (kapi.PodSpec, error) {
+	switch r := attr.GetObject().(type) {
+	case *kapi.Pod:
+		return r.Spec, nil
+	case *kapi.ReplicationController:
+		return r.Spec.Template.Spec, nil
+	case *extensions.Deployment:
+		return r.Spec.Template.Spec, nil
+	case *extensions.ReplicaSet:
+		return r.Spec.Template.Spec, nil
+	case *extensions.Job:
+		return r.Spec.Template.Spec, nil
+	case *deployapi.DeploymentConfig:
+		return r.Spec.Template.Spec, nil
+	}
+	return kapi.PodSpec{}, kapierrors.NewInternalError(fmt.Errorf("No PodSpec available for supplied admission attribute"))
+}
+
+// validate PodSpec if NodeName or NodeSelector are specified
+func (o *podNodeConstraints) admitPodSpec(attr admission.Attributes, ps kapi.PodSpec) error {
+	lbls := []string{}
+	// nodeSelector blacklist filter
+	if len(ps.NodeSelector) > 0 {
+		for nslbl := range ps.NodeSelector {
+			for _, bllbl := range o.config.NodeSelectorLabelBlacklist {
+				if bllbl == nslbl {
+					lbls = append(lbls, bllbl)
+				}
+			}
+		}
+	}
+	// nodeName constraint
+	if len(ps.NodeName) > 0 || len(lbls) > 0 {
+		allow, err := o.checkPodsBindAccess(attr)
+		if err != nil {
+			return err
+		}
+		if allow != nil && !allow.Allowed {
+			switch {
+			case len(ps.NodeName) > 0 && len(lbls) == 0:
+				return admission.NewForbidden(attr, fmt.Errorf("node selection by nodeName is prohibited by policy for your role"))
+			case len(ps.NodeName) == 0 && len(lbls) > 0:
+				return admission.NewForbidden(attr, fmt.Errorf("node selection by label(s) %v is prohibited by policy for your role", lbls))
+			case len(ps.NodeName) > 0 && len(lbls) > 0:
+				return admission.NewForbidden(attr, fmt.Errorf("node selection by nodeName and label(s) %v is prohibited by policy for your role", lbls))
+			}
+		}
+	}
+	return nil
+}
+
+func (o *podNodeConstraints) SetOpenshiftClient(c client.Interface) {
+	o.client = c
+}
+
+func (o *podNodeConstraints) Validate() error {
+	if o.client == nil {
+		return fmt.Errorf("PodNodeConstraints needs an Openshift client")
+	}
+	return nil
+}
+
+// build LocalSubjectAccessReview struct to validate role via checkAccess
+func (o *podNodeConstraints) checkPodsBindAccess(attr admission.Attributes) (*authorizationapi.SubjectAccessReviewResponse, error) {
```

Comment:
> @deads2k Would you be ok with passing the origin authorizer to build the Kube master? (Otherwise I don't see how we could pass one to the plugin when initialized by Kube)


@liggitt why do we have two separate admission chains anyway?